File: core/src/main/java/no/elg/infiniteBootleg/world/subgrid/contact/ContactManager.java
package no.elg.infiniteBootleg.world.subgrid.contact;

import static no.elg.infiniteBootleg.world.GlobalLockKt.BOX2D_LOCK;

import com.badlogic.ashley.core.Entity;
import com.badlogic.gdx.physics.box2d.Body;
import com.badlogic.gdx.physics.box2d.Contact;
import com.badlogic.gdx.physics.box2d.ContactImpulse;
import com.badlogic.gdx.physics.box2d.ContactListener;
import com.badlogic.gdx.physics.box2d.Manifold;
import no.elg.infiniteBootleg.world.box2d.CollisionListener;
import org.jetbrains.annotations.NotNull;

public class ContactManager implements ContactListener {

  private static void handleContact(@NotNull ContactType type, @NotNull Contact contact) {
    // Must be within a box2d lock to make sure the contact is valid
    synchronized (BOX2D_LOCK) {
      CollisionListener.INSTANCE.contact(type, contact);
      Body body = contact.getFixtureB().getBody();
      Object data = body.getUserData();
      if (data instanceof Entity) {
        CollisionListener.INSTANCE.contact(type, contact);
      }
      //      if (data instanceof Entity entity && !entity.isDisposed()) {
      //        entity.contact(type, contact);
      //      }
      //      if (data instanceof com.badlogic.ashley.core.Entity entity) {
      //        var engine = Main.inst().getEngine();
      //        if (engine == null) {
      //          Main.logger().warn("ContactManager", "Failed to get engine in handleContact");
      //          return;
      //        }
      //        CollisionListener.INSTANCE.contact(type, contact);
      //        handleAshleyContact(type, contact, engine, entity);
      //      }
    }
  }

  @Override
  public void beginContact(Contact contact) {
    handleContact(ContactType.BEGIN_CONTACT, contact);
  }

  @Override
  public void endContact(Contact contact) {
    handleContact(ContactType.END_CONTACT, contact);
  }

  @Override
  public void preSolve(Contact contact, Manifold oldManifold) {}

  @Override
  public void postSolve(Contact contact, ContactImpulse impulse) {}
}

File: core/src/main/java/no/elg/infiniteBootleg/world/subgrid/contact/ContactManager.kt
package no.elg.infiniteBootleg.world.subgrid.contact

import com.badlogic.ashley.core.Engine
import com.badlogic.gdx.physics.box2d.Contact
import com.badlogic.gdx.physics.box2d.ContactImpulse
import com.badlogic.gdx.physics.box2d.ContactListener
import com.badlogic.gdx.physics.box2d.Manifold
import no.elg.infiniteBootleg.world.ecs.components.events.ECSEvent.Companion.handleEvent
import no.elg.infiniteBootleg.world.ecs.components.events.PhysicsEvent

class ContactManager(val engine: Engine) : ContactListener {

  override fun beginContact(contact: Contact) = engine.handleEvent(PhysicsEvent.ContactBeginsEvent(contact)) // { contact.containsEntityFixture(it) }
  override fun endContact(contact: Contact) = engine.handleEvent(PhysicsEvent.ContactEndsEvent(contact)) // { contact.containsEntityFixture(it) }
  override fun preSolve(contact: Contact, oldManifold: Manifold) =
    engine.handleEvent(PhysicsEvent.PreSolveContactEvent(contact, oldManifold)) // { contact.containsEntityFixture(it) }

  override fun postSolve(contact: Contact, impulse: ContactImpulse) =
    engine.handleEvent(PhysicsEvent.PostSolveContactEvent(contact, impulse)) // { contact.containsEntityFixture(it) }
}
